File: Graph/DFS_base_code.py
class GNode:
    def __init__(self, id, c = "W", d = -1, f = -1, p = None):
        self.id = id
        self.color = c 
        self.dist = d 
        self.fin = f 
        self.parent = p 
        
    def __str__(self):
        if (self.parent != None):
            return "(" + self.id + ", " + self.color + ", (" \
                + str(self.dist) + ", " + str(self.fin) + "), " + self.parent.id + ")"
        else:
            return "(" + self.id + ", " + self.color + ", (" \
                + str(self.dist) + ", " + str(self.fin) + "), None)"

def printG(g):
    print("====================================================")
    for key, val in g.items():
        print(str(key) + ": ", end="")
        for u in val:
            print(str(u), end = " ")
        print("")

time = 0
def DFS(G, st):
    # 시작점 방문표시 
    global time 
    time += 1
    st.dist = time 
    st.color = "G"
    printG(G)
    for nxt in G[st]:
        if (nxt.color == "W"):
            nxt.parent = st
            DFS(G, nxt)
    nxt.color = "B"
    time += 1 
    st.fin = time
    printG(G)

u, v, w = GNode("u"), GNode("v"), GNode("w")
x, y, z = GNode("x"), GNode("y"), GNode("z")
# Directed adj
G = dict()
G[u] = [v, x]   # u -> v, x 
G[v] = [y]
G[w] = [y, z]
G[x] = [v]
G[y] = [x]
G[z] = [z]

for vertex in G:
    if (vertex.color == "W"):
        DFS(G, vertex)

File: Graph/DFS_base_code.cpp
#include <iostream>
#include <vector>
#include <map>

class GNode {
public:
    std::string id;
    std::string color;
    int dist;
    int fin;
    GNode* parent;

    GNode(std::string _id, std::string c = "W", int d = -1, int f = -1, GNode* p = nullptr)
        : id(_id), color(c), dist(d), fin(f), parent(p) {}

    std::string toString() {
        if (parent != nullptr) {
            return "(" + id + ", " + color + ", (" + std::to_string(dist) + ", " + std::to_string(fin) + "), " + parent->id + ")";
        } else {
            return "(" + id + ", " + color + ", (" + std::to_string(dist) + ", " + std::to_string(fin) + "), None)";
        }
    }
};

void printG(std::map<GNode*, std::vector<GNode*>>& g) {
    std::cout << "====================================================" << std::endl;
    for (const auto& entry : g) {
        std::cout << entry.first->id << ": ";
        for (const GNode* u : entry.second) {
            std::cout << u->id << " ";
        }
        std::cout << std::endl;
    }
}

int time = 0;

void DFS(std::map<GNode*, std::vector<GNode*>>& G, GNode* st) {
    // 시작점 방문표시
    time += 1;
    st->dist = time;
    st->color = "G";
    printG(G);
    for (GNode* nxt : G[st]) {
        if (nxt->color == "W") {
            nxt->parent = st;
            DFS(G, nxt);
        }
    }
    st->color = "B";
    time += 1;
    st->fin = time;
    printG(G);
}

int main() {
    GNode u("u"), v("v"), w("w");
    GNode x("x"), y("y"), z("z");

    std::map<GNode*, std::vector<GNode*>> G;
    G[&u] = {&v, &x};
    G[&v] = {&y};
    G[&w] = {&y, &z};
    G[&x] = {&v};
    G[&y] = {&x};
    G[&z] = {&z};

    for (auto& vertex : G) {
        if (vertex.first->color == "W") {
            DFS(G, vertex.first);
        }
    }

    return 0;
}
